src/Tasks/DownloadFile.cs

```diff
@@ -0,0 +1,280 @@
+        private bool IsRetriable(Exception exception, out Exception actualException)
+        {
+            actualException = exception;
+
+            // Get aggregate inner exception
+            if (actualException is AggregateException aggregateException && aggregateException.InnerException != null)
+            {
+                actualException = aggregateException.InnerException;
+            }
+
+            // Some HttpRequestException have an inner exception that has the real error
+            if (actualException is HttpRequestException httpRequestException && httpRequestException.InnerException != null)
+            {
+                actualException = httpRequestException.InnerException;
+
+                // An IOException inside of a HttpRequestException means that something went wrong while downloading
+                if (actualException is IOException)
+                {
+                    return true;
+                }
+            }
+
+            if (actualException is CustomHttpRequestException customHttpRequestException)
+            {
+                // A wrapped CustomHttpRequestException has the status code from the error
+                switch (customHttpRequestException.StatusCode)
+                {
+                    case HttpStatusCode.InternalServerError:
+                    case HttpStatusCode.RequestTimeout:
+                        return true;
+                }
+            }
+
+            if (actualException is WebException webException)
+            {
+                // WebException is thrown when accessing the Content of the response
+                switch (webException.Status)
+                {
+                    // Don't retry on anything that cannot be compensated for
+                    case WebExceptionStatus.TrustFailure:
+                    case WebExceptionStatus.MessageLengthLimitExceeded:
+                    case WebExceptionStatus.RequestProhibitedByCachePolicy:
+                    case WebExceptionStatus.RequestProhibitedByProxy:
+                        return false;
+
+                    default:
+                        // Retry on all other WebExceptions
+                        return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Attempts to get the file name to use when downloading the file.
+        /// </summary>
+        /// <param name="uri">The <see cref="Uri"/>.</param>
+        /// <param name="filename">Receives the name of the file.</param>
+        /// <returns><code>true</code> if a file name could be determined, otherwise <code>false</code>.</returns>
+        private bool TryGetFileName(Uri uri, out string filename)
+        {
+            // Not all URIs contain a file name so users will have to specify one
+            // Example: http://www.download.com/file/1/
+
+            filename = !String.IsNullOrWhiteSpace(DestinationFileName?.ItemSpec)
```

Comment:
Could we call TrygetFileName after we have recieved the headers?

Precedence (descending order):

* `DestinationFileName` (user-specified)
* `client.ResponseHeaders["content-disposition"]`, if `DestinationFileName` is empty
* `url.LocalPath`, if `DestinationFileName` is empty and `client.ResponseHeaders["content-disposition"]` is non-empty and meaningful (`<file-name>.<extension>` format)

so we can call urls like https://lorempizza.com/380/240, we get this kind of filename from response header: `imagedoc-darknoise.png` (instead of 240)